### DISCUSSION TITLE
Cursor Seems to be hidden under certain text in NVIM

### DISCUSSION THREAD

**Discussion by Ducky6944** [wezterm/wezterm#1830] [Q&A]
I wanted to make sure I didn't miss a configuration option, and check to see if this is a wezterm issue or a nvim/nvim plugin issue. I am unable to reproduce this behavior in kitty. I know NVIM uses some kind of conceal and virtual text system but I haven't dived too deep in that yet. 

In this picture you can see a Block Outline Cursor that has it's left side hidden by an indent guideline. 
<img width="2056" alt="Screen Shot 2022-04-07 at 12 07 33 AM" src="https://user-images.githubusercontent.com/12100960/162118781-5c7e6622-ad5a-4509-828a-5b8cc4e52a97.png">

Here you can('t) see the cursor hidden by the same guide line in insert mode. 
<img width="2056" alt="Screen Shot 2022-04-07 at 12 15 33 AM" src="https://user-images.githubusercontent.com/12100960/162119092-135a274b-e8d9-448b-afb4-c5ef87c281ba.png">

Kitty Referencing the first and second photos respectively
<img width="2056" alt="Screen Shot 2022-04-07 at 12 19 31 AM" src="https://user-images.githubusercontent.com/12100960/162119390-884f3f68-f6eb-4eb5-8b31-18cfab007240.png">

<img width="2056" alt="Screen Shot 2022-04-07 at 12 19 34 AM" src="https://user-images.githubusercontent.com/12100960/162119401-93645cbe-bf58-460d-b545-f337750efbfb.png">

Configuration
Nvim 
```
NVIM v0.7.0-dev+1383-g2a4660014
Build type: Release
LuaJIT 2.1.0-beta3
```
NVIM Plugin that seems to cause this effect
```
https://github.com/lukas-reineke/indent-blankline.nvim
```
Computer
```
macOS Monterey 12.1 (21C52)
M1 Max
```

<details>
  <summary>wezterm.lua</summary>

```LUA
local wezterm = require 'wezterm'
return {
  -- Colors
  color_scheme_dirs = {"$HOME/.config/wezterm/colors"},
  color_scheme = "duskfox",

  -- Fonts
  font = wezterm.font('VictorMono Nerd Font'),
  font_size = 18,

  -- Window
  window_decorations = "RESIZE",
  skip_close_confirmation_for_processes_named = {
    "bash", "sh", "zsh", "fish", "tmux"
  },
  hide_tab_bar_if_only_one_tab = true,
  window_padding = {
    left = "0.35cell",
    right = "0.5cell",
    top = "0.15cell",
    bottom = "0cell",
  },
  adjust_window_size_when_changing_font_size = false,
  use_fancy_tab_bar = false,
  colors = {
    -- NOTE: Fancy Tab Has it's own config
    -- NOTE: Should Dynamically pull values so it updates with themes
    tab_bar = {
      background = "#232136",
      active_tab = {
        bg_color = "#47407d",
        fg_color = "#e0def4"
      },
      inactive_tab = {
        bg_color = "#393552",
        fg_color = "#e0def4"
      },
      inactive_tab_hover = {
        bg_color = "#3b3052",
        fg_color = "#909090",
        italic = true,
      },
      new_tab = {
        bg_color = "#1b1032",
        fg_color = "#808080",
      },
      new_tab_hover = {
        bg_color = "#3b3052",
        fg_color = "#909090",
        italic = true,
      },
    },
  },

  -- Cursor
  default_cursor_style = "BlinkingBar",
  cursor_blink_rate = 400,
  cursor_blink_ease_in = {CubicBezier={0.0, 0.0, 0.0, 0.0}},
  cursor_blink_ease_out = {CubicBezier={0.0, 0.0, 0.0, 0.0}},

  -- Key Mappings
  keys = {
    -- {key="LeftArrow", mods="OPT", action=wezterm.action{SendString="\x1bb"}},
    -- ALT
    -- TODO: It may be better to change this to a string for better control in other programs
    {key="LeftArrow", mods="OPT", action={SendKey={key="b", mods="ALT"}}},
    {key="RightArrow", mods="OPT", action={SendKey={key="f", mods="ALT"}}},

    -- CMD
    -- TODO: CMD+SHIFT+UP/DOWN
    {key="LeftArrow", mods="CMD", action=wezterm.action{SendString="☀︎"}},
    {key="LeftArrow", mods="CMD|SHIFT", action=wezterm.action{SendString="☍"}},
    {key="RightArrow", mods="CMD", action=wezterm.action{SendString="☼"}},
    {key="RightArrow", mods="CMD|SHIFT", action=wezterm.action{SendString="☌"}},
    {key="UpArrow", mods="CMD", action=wezterm.action{SendString="☔︎"}},
    {key="DownArrow", mods="CMD", action=wezterm.action{SendString="☂︎"}},
    {key="a", mods="CMD", action=wezterm.action{SendString="♗"}},
    {key="z", mods="CMD", action=wezterm.action{SendString="☾"}},
    {key="z", mods="CMD|SHIFT", action=wezterm.action{SendString="☁︎"}},
    {key="Backspace", mods="CMD", action=wezterm.action{SendString="☽"}},
    {key="/", mods="CMD", action=wezterm.action{SendString="♇"}},

    -- Set Env Variables for Spawned Programs
    set_environment_variables = {
      TEST = "TEST"
    }
  },
}

```

</details>

**Comment by wez** [comment]
While those screenshots look beautiful, navigating those giant images on the displays I have available is difficult and it's hard to compare them!  Would you mind making more tightly focused screenshots?

How are those guidelines produced in nvim? Are they box drawing glyphs or does it use an image protocol to put those on the screen?

**Reply by Ducky6944** [reply]
So sorry about that. Here's a couple of screenshots that are a little more focused and hopefully easier to see. 

<img width="170" alt="Screen Shot 2022-04-07 at 10 47 05 AM" src="https://user-images.githubusercontent.com/12100960/162226730-dfb75e53-3a05-41f7-a7d5-8276ed66be98.png">
<img width="126" alt="Screen Shot 2022-04-07 at 10 48 02 AM" src="https://user-images.githubusercontent.com/12100960/162226943-42308044-d4a9-4b4a-91fb-ed7708cb0cd6.png">

Looking through their docs, it looks like they use a feature called [virtual text](https://github.com/lukas-reineke/indent-blankline.nvim/blob/cd5b800ed9545a944fc3b5bd55ef12da90944e7e/doc/indent_blankline.txt#L23). The configuration for [indent-blankline](https://github.com/lukas-reineke/indent-blankline.nvim/blob/cd5b800ed9545a944fc3b5bd55ef12da90944e7e/doc/indent_blankline.txt#L132) accepts a character. By default it uses ` let g:indent_blankline_char = '|'`.  

Some information about this character
```
▏
LEFT ONE EIGHTH BLOCK
Unicode: U+258F, UTF-8: E2 96 8F
```

You can see that the glyph cuts through the center of the cursor using the character below. 
```
┃
BOX DRAWINGS HEAVY VERTICAL
Unicode: U+2503, UTF-8: E2 94 83
```

<img width="187" alt="Screen Shot 2022-04-07 at 11 01 48 AM" src="https://user-images.githubusercontent.com/12100960/162230149-2b1c052e-d987-4919-92b3-a12fca53801e.png">

It also seems that when the guideline is active and highlighted that the cursor has no issue being show as you can see below. 

<img width="143" alt="Screen Shot 2022-04-07 at 10 55 03 AM" src="https://user-images.githubusercontent.com/12100960/162228603-40281476-2380-4190-900d-c0dae530ce1b.png">
<img width="146" alt="Screen Shot 2022-04-07 at 10 54 56 AM" src="https://user-images.githubusercontent.com/12100960/162228628-b04facc5-d10e-4d05-bcb6-2cc45e685836.png">

Here is a possibly relevant [PR](https://github.com/neovim/neovim/pull/8180) regarding it. 
Also a couple of possibly relavent api calls from the manual:
- *nvim_buf_set_extmark()
- *nvim_buf_set_virtual_text() -- Deprecated to above

**Reply by wez** [reply]
The cursor is implemented as a graphic that is rendered on the background layer, so that it never obscures the text content of a cell.
When the cursor intersects with a cell, the fg of the cell is changed according to the cursor_fg color in your color scheme.

This is an interesting (literal!) edge case: we guarantee not to hide the text that you wrote, but that text can obscure certain cursor styles. If we made the cursor win, you'd not be able to see that particular glyph when you enter it directly and cursor over it.

I don't think there's a clear winner for this, except perhaps to enable blinking and then you'd be able to see it.

**Reply by Ducky6944** [reply]
That's interesting. I have blinking enabled. Maybe it needs to be enabled in neovim as well. I feel like the cursor should win personally, just for usability's sake, but I'm sure this is a more profound decision than I may be considering. Is this something a novice like me could patch and build or maybe toggle? Totally understand if this is not worth the added complexity of supporting though.

**Comment by Ducky6944** [comment]
@wez This seems to be fixed at some point in the last year-ish. Not sure if Neovim, the plugin, or you did something, but I just thought I'd comment on it for posterity.
